packages/react-components/react-swatch-picker-preview/src/utils/renderUtils.tsx

```diff
@@ -1,31 +1,34 @@
 import * as React from 'react';
-import { ColorSwatch, ColorSwatchProps, SwatchPickerRow } from '../';
+import { ColorSwatch, SwatchPickerRow, ImageSwatch } from '../';
+import type { ColorSwatchProps, ImageSwatchProps } from '../';
 
-export const renderSwatchPickerRow = (
-  colors: ColorSwatchProps[],
-  rowIdx: number,
-  SwatchElement: React.ElementType = ColorSwatch,
-) => {
-  return (
-    <SwatchPickerRow key={`row-${rowIdx}`}>
-      {colors.map(color => (
-        <SwatchElement key={color.value} {...color} />
-      ))}
-    </SwatchPickerRow>
-  );
+export type SwatchProps = ImageSwatchProps | ColorSwatchProps;
+
+export type SwatchPickerGridProps = {
+  items: SwatchProps[];
+  columnCount: number;
+  renderRow?: (props: { children: JSX.Element[]; rowId: string | number }) => JSX.Element;
+  renderSwatch?: (item: SwatchProps) => JSX.Element;
 };
 
-export const renderSwatchPickerGrid = (
-  colors: ColorSwatchProps[],
-  columnCount: number,
-  renderRow: (colors: ColorSwatchProps[], rowIdx: number) => JSX.Element = renderSwatchPickerRow,
-) => {
-  const rowCount = Math.ceil(colors.length / columnCount);
+export const renderSwatchPickerGrid = (props: SwatchPickerGridProps) => {
+  const { items, columnCount, renderRow, renderSwatch } = props;
+  const _renderRow = renderRow || (({ children, rowId }) => <SwatchPickerRow key={rowId}>{children}</SwatchPickerRow>);
+  const _renderSwatch =
+    renderSwatch ||
+    ((item: SwatchProps) =>
+      (item as ImageSwatchProps).src ? (
+        <ImageSwatch key={item.value} src={(item as ImageSwatchProps).src ?? ''} {...item} />
+      ) : (
+        <ColorSwatch key={item.value} color={item.color || ''} {...item} />
+      ));
+
+  const rowCount = Math.ceil(items.length / columnCount);
   const rows = Array.from({ length: rowCount }, (_, i) => {
     const start = i * columnCount;
     const end = start + columnCount;
-    return colors.slice(start, end);
+    return items.slice(start, end);
   });
 
-  return rows.map(renderRow);
+  return rows.map((row, index) => _renderRow({ children: row.map(_renderSwatch), rowId: index }));
 };
```

Comment:
```ts
function isImageSwatchProps(item: SwatchProps): item is ImageSwatchProps {
  return (item as ImageSwatchProps).src !== undefined;
}

const defaultRenderRow: NonNullable<SwatchPickerGridProps['renderRow']> = ({ children, rowId }) => (
  <SwatchPickerRow key={rowId}>{children}</SwatchPickerRow>
);

const defaultRenderSwatch = (item: SwatchProps) =>
  isImageSwatchProps(item) ? <ImageSwatch {...item} key={item.value} /> : <ColorSwatch {...item} key={item.value} />;

export const renderSwatchPickerGrid = (props: SwatchPickerGridProps) => {
  const { items, columnCount, renderRow = defaultRenderRow, renderSwatch = defaultRenderSwatch } = props;

  const rowCount = Math.ceil(items.length / columnCount);
  const rows = Array.from({ length: rowCount }, (_, i) => {
    const start = i * columnCount;
    const end = start + columnCount;
    return items.slice(start, end);
  });

  return rows.map((row, index) => renderRow({ children: row.map(renderSwatch), rowId: index }));
};
```

What about these changes? 🐱 I think that we will need to find a better way to distinguish between `Image` & `Color`, but let's leave it as it is for now